### PR TITLE
feat(drivers): Enable automatic driver management for Firefox

### PR DESCRIPTION
**Pull request by nextlevelbeard**
## Proposed changes
Enables automatic driver management for Firefox and its channels.

Fixes https://github.com/webdriverio/webdriverio/issues/10924

Allows `browserVersion` capability to be set to a Firefox release channel:

i.e. `'stable' | 'latest' | 'beta' | 'dev' | 'esr'`

As well as any published version.

Example capabilities:
```js
{
	browserName: 'firefox',
	browserVersion: 'nightly' // Latest Nightly
},
{
	browserName: 'firefox',
	browserVersion: '118.0a1' // Specific Nightly
},
{
	browserName: 'firefox',
	browserVersion: 'esr' // Latest Extended Support Release
},
{
	browserName: 'firefox',
	browserVersion: '115.1.0esr' // Specifc Extended Support Release
},
{
	browserName: 'firefox',
	browserVersion: 'beta' // Latest Beta
},
{
	browserName: 'firefox',
	browserVersion: '117.0b2' // Specific Beta (non Developer Edition)
},
{
	browserName: 'firefox',
	browserVersion: 'latest' // sames as stable, latest released
},
{
	browserName: 'firefox',
	browserVersion: '114.0.1' // Specific past stable version
},
{
	browserName: 'firefox',
	browserVersion: 'dev' // Latest Developer Edition
},
{
	browserName: 'firefox',
	browserVersion: '117.0b2dev' // Specific Developer Edition
}
```

Please test more thoroughly on Windows and Linux.

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by nextlevelbeard** [comment]
Needs more tests to cover global threshold.
Mind testing / helping @christian-bromann ?

File: packages/wdio-utils/src/driver/index.ts [hunk 1]
```diff
@@ -106,9 +108,21 @@ export async function startWebDriver (options: Options.WebDriver) {
         /**
          * Firefox
          */
-        const geckodriverOptions = caps['wdio:geckodriverOptions'] || ({} as GeckodriverParameters)
+        const geckodriverOptions = caps['wdio:geckodriverOptions'] || {}
+        const cacheDir = geckodriverOptions.cacheDir || options.cacheDir || os.tmpdir()
```
Comment (christian-bromann) [inline packages/wdio-utils/src/driver/index.ts:112]: We should reuse `cacheDir` from line `73`

File: packages/wdio-utils/src/driver/utils.ts [hunk 2]
```diff
@@ -1,15 +1,21 @@
-import os from 'node:os'
+import os, { platform } from 'node:os'
```
Comment (christian-bromann) [inline packages/wdio-utils/src/driver/utils.ts:1]: just use `os.platform` or `process.platform`

File: packages/wdio-utils/src/driver/utils.ts [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
 import fs from 'node:fs'
 import fsp from 'node:fs/promises'
+import { mkdir, readdir, unlink } from 'node:fs/promises'
```
Comment (christian-bromann) [inline packages/wdio-utils/src/driver/utils.ts:4]: just use `fsp.mkdir` etc.

File: packages/wdio-utils/src/driver/utils.ts [hunk 4]
```diff
@@ -19,9 +25,88 @@ import type { Options, Capabilities } from '@wdio/types'
 
 import { DEFAULT_HOSTNAME, DEFAULT_PROTOCOL, DEFAULT_PATH, SUPPORTED_BROWSERNAMES } from '../constants.js'
 
+import extractZip from 'extract-zip'
+import tar from 'tar-fs'
+import bzip from 'unbzip2-stream'
```
Comment (christian-bromann) [inline packages/wdio-utils/src/driver/utils.ts:30]: let's group these imports with other module imports at the top of the file.

File: packages/wdio-utils/src/driver/utils.ts [hunk 5]
```diff
@@ -125,7 +210,289 @@ export async function setupChrome(cacheDir: string, caps: Capabilities.Capabilit
...
+    await unpackArchive(archivePath, outputPath)
+
+    if (fs.existsSync(archivePath)) {
+        await unlink(archivePath).catch(err => log.error(err))
+    }
+
+    const executablePath = getExecPath(OS, outputPath)
+
+    return { executablePath, buildId: channel, platform: arch }
+}
```
Comment (christian-bromann) [inline packages/wdio-utils/src/driver/utils.ts:493]: maybe we should move all browser related code into a `chrome.ts` or `firefox.ts`

File: packages/wdio-utils/tests/driver/utils.test.ts [hunk 6]
```diff
@@ -30,30 +46,76 @@ vi.mock('node:fs', () => ({
...
+        default: {
+            ...origFS,
+            ...mod,
+        }
+    }
+})
+
+vi.mock('node:child_process', async () => {
+
+    const origCP = await vi.importActual<typeof import('node:child_process')>('node:child_process')
```
Comment (christian-bromann) [inline packages/wdio-utils/tests/driver/utils.test.ts:81]: I honestly would not mix actual implementation an mock together. We should just enhance the mock file rather.

**Review by christian-bromann — COMMENTED** [review]
Some initial comments

File: packages/wdio-utils/src/driver/utils.ts (hunk 5 above)
Comment (nextlevelbeard) [reply packages/wdio-utils/src/driver/utils.ts]: done

**Comment by christian-bromann** [comment]
@nextlevelbeard you might want to check the Puppeteer project and see if they will provide the same download capabilities for Firefox as for Chrome. I am seeing [some signs](https://github.com/puppeteer/puppeteer/blob/ce9e2df04d55829f9cdca273513e58c087b04667/packages/browsers/src/install.ts#L87-L88) that they will do so which would allow use to re-use a lot of code and don't need to manually implement it.

**Comment by christian-bromann** [comment]
@nextlevelbeard I [looked into this](https://github.com/webdriverio/webdriverio/pull/10995) and it seems like `@puppeteer/browser` does most of the job for us. However it doesn't support downloading specific Firefox versions. I'ld suggest to build in top of that PR and update the `setupPuppeteerBrowser` at this point:

```ts
    const isCombinationAvailable = await canDownload(installOptions)
    if (!isCombinationAvailable) {
        throw new Error(`Couldn't find a matching Chrome browser for tag "${buildId}" on platform "${platform}"`)
    }
```

Here, if we provide a Firefox version, it will fail since `@puppeteer/browser` doesn't support it. It only supports `latest` as browser version which will download Nightly. I would suggest to check whether we deal with Firefox here and have it download that specific version.

**Comment by nextlevelbeard** [comment]
@christian-bromann  I originally looked into using more of `@puppeteer/browsers` but found many limitations like you mentioned, but I see your point and more of it can be used, will work over `setupPuppeteerBrowser` and https://github.com/webdriverio/webdriverio/pull/10973 💪

**Comment by christian-bromann** [comment]
I will go ahead and close the PR given the amount of conflicts. We should follow what happens in https://github.com/puppeteer/puppeteer/issues/10420 as it seems we soon have this supported out of the box.

**Pull request closed without merging**